src/options/parser.rs

```diff
@@ -1,815 +1,287 @@
-//! A general parser for command-line options.
-//!
-//! exa uses its own hand-rolled parser for command-line options. It supports
-//! the following syntax:
-//!
-//! - Long options: `--inode`, `--grid`
-//! - Long options with values: `--sort size`, `--level=4`
-//! - Short options: `-i`, `-G`
-//! - Short options with values: `-ssize`, `-L=4`
-//!
-//! These values can be mixed and matched: `exa -lssize --grid`. If you’ve used
-//! other command-line programs, then hopefully it’ll work much like them.
-//!
-//! Because exa already has its own files for the help text, shell completions,
-//! man page, and readme, so it can get away with having the options parser do
-//! very little: all it really needs to do is parse a slice of strings.
-//!
-//!
-//! ## UTF-8 and `OsStr`
-//!
-//! The parser uses `OsStr` as its string type. This is necessary for exa to
-//! list files that have invalid UTF-8 in their names: by treating file paths
-//! as bytes with no encoding, a file can be specified on the command-line and
-//! be looked up without having to be encoded into a `str` first.
-//!
-//! It also avoids the overhead of checking for invalid UTF-8 when parsing
-//! command-line options, as all the options and their values (such as
-//! `--sort size`) are guaranteed to just be 8-bit ASCII.
-
-use std::ffi::{OsStr, OsString};
-use std::fmt;
-
-use crate::options::error::{Choices, OptionsError};
-
-/// A **short argument** is a single ASCII character.
-pub type ShortArg = u8;
-
-/// A **long argument** is a string. This can be a UTF-8 string, even though
-/// the arguments will all be unchecked `OsString` values, because we don’t
-/// actually store the user’s input after it’s been matched to a flag, we just
-/// store which flag it was.
-pub type LongArg = &'static str;
-
-/// A **list of values** that an option can have, to be displayed when the
-/// user enters an invalid one or skips it.
-///
-/// This is literally just help text, and won’t be used to validate a value to
-/// see if it’s correct.
-pub type Values = &'static [&'static str];
-
-/// A **flag** is either of the two argument types, because they have to
-/// be in the same array together.
-#[derive(PartialEq, Eq, Debug, Copy, Clone)]
-pub enum Flag {
-    Short(ShortArg),
-    Long(LongArg),
-}
-
-impl Flag {
-    pub fn matches(&self, arg: &Arg) -> bool {
-        match self {
-            Self::Short(short) => arg.short == Some(*short),
-            Self::Long(long) => arg.long == *long,
-        }
-    }
+pub use clap::Parser;
+use clap::ValueEnum;
+use std::ffi::OsString;
+
+#[derive(Parser)]
+#[command(author, version, about, long_about)] // Read from `Cargo.toml`
+#[clap(disable_help_flag = true)]
+pub struct Opts {
+    pub paths: Vec<OsString>,
+    /// Show hidden files.
+    #[arg(short, long, action = clap::ArgAction::Count)]
```

Comment:
I think I made a comment about this before but maybe I forgot to click the "submit" button because I can't find it anymore. I assume the `ArgAction::Count` is to allow multiple occurrences right? You could get the same behaviour with `#[command(args_override_self = true)]`

See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b76648135d08a519d66e46ae51d0d63b

Another thing to think about is which argument "wins" if they conflict. As far as I can tell, you don't really have any logic for that. Consider for example `eza -GT` and `eza -TG`.

Reply:
I don't think `clap` allows `arg=true` by default for boolean fields, though? The default `ArgAction` for `bool` is `SetTrue`, which doesn't take an argument. Do you have an example of an argument being verified by count? I'm not entirely sure what you mean by that.

Reply:
all is verified by count as `-a` and `-a -a` is different, so I thought lets put everything under count in order to be ready if anything like that is tb addeed

Reply:
Right, that makes sense. I still disagree, because I think that the distinction makes the intent of the argument more clear. For example, when you look at a field, you can't see for which arguments the count matters and for which it doesn't. But this is exactly where I defer to the maintainers here for a decision. They might agree with you :)

Reply:
Well I implemented to match how the original parser was made but still making it kinda easy to understand